File: spring-boot/src/main/java/com/SuperNova/service/CourseService.java
package com.SuperNova.service;
import com.SuperNova.model.Course;
import com.SuperNova.core.Service;

import java.util.List;


/**
 * Created by Chongli on 2020/05/14.
 */
public interface CourseService extends Service<Course> {

    /**
     * 获取个人的课程(分页)
     * @param u_id
     * @return
     */
    String getMyCourses(String u_id,int pageIndex,int pageSize);

    /**
     * 更新课程相关信息
     * @param course
     */
    void updateCourse(Course course);

    /**
     * 添加课程
     * @param course
     * @return 返回课程id
     */
    int addCourse(Course course);

    /**
     * 学生选课
     * @param c_id
     * @param u_id
     */
    void joinCourse(int c_id,String u_id);

    /**
     * 获取自己的其它未选的课程/未教的课程(已发布)+教师信息 (分页)
     * @param u_id
     * @return
     */
    String searchOtherCourses(String u_id,int pageIndex,int pageSize);

    /**
     * 修改课程状态(删除：-1,未发布：0,已发布：1)
     * @param c_id
     * @param status
     */
    void changeCourseStatus(int c_id,String status);

    /**
     * 获取所有课程(分页)
     * @return
     */
    String searchAllCourses(int pageIndex,int pageSize);

    /**
     *
     * @param u_id
     * @return
     */
    String searchAllMyCourses(String u_id);

    /**
     * 判断该用户是否为该课程的主讲老师
     * @param u_id
     * @param c_id
     * @return
     */
    boolean isTeacher(String u_id,int c_id);
}

File: supernova/src/app/services/course.service.ts
import { Injectable } from '@angular/core';
import {
  HttpClient,
  HttpParams,
  HttpHeaders,
  HttpRequest,
  HttpResponse,
} from '@angular/common/http';
import { Response } from '../models/generic-response';
import { Course } from '../models/course';
import { User } from '../models/user';
import { UserService } from './user.service';
import { environment } from 'src/environments/environment';
import { filter } from 'rxjs/operators';

@Injectable({
  providedIn: 'root',
})
export class CourseService {
  requestURL = {
    my: `${environment.apiUrl}:8026/coursesController/searchMyCourses`,
    other: `${environment.apiUrl}:8026/coursesController/searchOtherCourses`,
    all: `${environment.apiUrl}:8026/coursesController/searchAllCourses`,
    all_my: `${environment.apiUrl}:8026/coursesController/searchAllMyCourses`,
  };

  constructor(private http: HttpClient, private userService: UserService) {}

  getCourse(c_id: number) {
    const params = new HttpParams({
      fromObject: {
        pbl_token: String(this.userService.getUser().token),
        c_id: String(c_id),
      },
    });
    return this.http.get<Response<{ course: Course }>>(
      `${environment.apiUrl}:8026/coursesController/searchCourseByCid`,
      { params }
    );
  }

  getCourses(
    type: string,
    userId: string,
    pageIndex: number,
    pageSize: number
  ) {
    const params = new HttpParams({
      fromObject: {
        pbl_token: String(this.userService.getUser().token),
        u_id: userId,
        pageIndex: String(pageIndex),
        pageSize: String(pageSize),
      },
    });
    return this.http.get<
      Response<{
        courses: {
          list: Course[];
        };
        teachers: {
          list: User[];
        };
        total: number;
      }>
    >(this.requestURL[type], { params });
  }

  getMyCourseNames() {
    const params = new HttpParams({
      fromObject: {
        pbl_token: String(this.userService.getUser().token),
      },
    });
    return this.http.get<Response<{ courses: Course[]; type: string }>>(
      this.requestURL.all_my,
      { params }
    );
  }

  joinCourse(courseId: number) {
    const headers = new HttpHeaders({
      'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8',
    });
    const params = new HttpParams({
      fromObject: {
        pbl_token: String(this.userService.getUser().token),
        s_id: String(this.userService.getUser().u_id),
        c_id: String(courseId),
      },
    });
    return this.http.post<Response<{}>>(
      `${environment.apiUrl}:8026/coursesController/joinCourse`,
      params.toString(),
      {
        headers,
      }
    );
  }

  createCourse(course: Course, image: any) {
    const params = new HttpParams({
      fromObject: {
        pbl_token: String(JSON.parse(localStorage.getItem('User')).token),
        course: JSON.stringify(course),
      },
    });
    const formData = new FormData();
    formData.append('image', image);
    const headers = new HttpHeaders({
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Methods': 'POST,GET,PUT',
      'Access-Control-Max-Age': '1728000',
      'Access-Control-Allow-Headers':
        'Content-Type,Access-Token,Authorization,ybg',
    });
    const req = new HttpRequest(
      'POST',
      `${environment.apiUrl}:8026/coursesController/addCourse`,
      formData,
      {
        headers,
        params,
      }
    );
    return this.http
      .request(req)
      .pipe(filter((e) => e instanceof HttpResponse));
  }

  changeCourse(course: Course) {
    const params = new HttpParams({
      fromObject: {
        pbl_token: String(this.userService.getUser().token),
        course: JSON.stringify(course),
      },
    });
    return this.http.put<Response<{}>>(
      `${environment.apiUrl}:8026/coursesController/changeCourse`,
      params
    );
  }

  changeCourseWithImg(course: Course, image: any) {
    const params = new HttpParams({
      fromObject: {
        pbl_token: String(JSON.parse(localStorage.getItem('User')).token),
        course: JSON.stringify(course),
      },
    });
    const formData = new FormData();
    formData.append('image', image);
    const headers = new HttpHeaders({
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Methods': 'POST,GET,PUT',
      'Access-Control-Max-Age': '1728000',
      'Access-Control-Allow-Headers':
        'Content-Type,Access-Token,Authorization,ybg',
    });
    const req = new HttpRequest(
      'POST',
      `${environment.apiUrl}:8026/coursesController/changeCourseWithImg`,
      formData,
      {
        headers,
        params,
      }
    );
    return this.http
      .request(req)
      .pipe(filter((e) => e instanceof HttpResponse));
  }

  deleteCourse(course: Course) {
    course.status = 0;
    return this.changeCourse(course);
  }

  publishCourse(course: Course) {
    course.status = 2;
    return this.changeCourse(course);
  }

  resumeCourse(course: Course) {
    return this.publishCourse(course);
  }
}
